Add legacySegments for load balance segments rc track

congqixia opened the pull request:
- Add legacySegments in ShardCluster tracking old segment inUse
- Apply node id check for segment online/inUse check
- Add unit test simulating load balance case

Fix #16830 

/kind bug

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

letian-jiang commented on internal/querynode/shard_cluster.go:
```diff
@@ -303,6 +308,17 @@ func (sc *ShardCluster) transferSegment(old *shardSegmentInfo, evt segmentEvent)
+				inUse:     old.inUse,
+				state:     old.state,
+			})
+			// set segment inUse to zero, since there is no allocation for new segment
```
I think the segment here refers to the 'old' segment, whose inUse counter is moved to the legacy segment.

congqixia committed:
Add legacySegments for load balance segments rc track

- Add legacySegments in ShardCluster tracking old segment inUse
- Apply node id check for segment online/inUse check
- Add unit test simulating load balance case

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

letian-jiang commented:
/lgtm
